CIP-????/README.md

```diff
@@ -0,0 +1,294 @@
+by 2035. However, the smart contracts have long development and lifetime,
+so migration to post-quantum cryptography algorithms should start several
+years before they would become insecure.
+This summer NIST has recommended three PQC signature algorithms
+for standardization, publishing their reference implementations,
+and finishing initial comment period.
+
+Supporting PQC signatures will bring the following benefits:
+
+* smart contracts may use future standard algorithms today;
+* early support for PQC will simplify future PQC migration, if needed;
+* future PQC migration would be smooth and require less changes;
+* support for PQC may increase industry acceptance for long running projects;
+* users that want their wallets to survive over ten years may
+  use smart contracts verifying multiple algorithms today;
+* Cardano will become first blockchain to verify PQC standard signatures,
+  thus getting first-mover advantage.
+
+## Use cases
+
+Adding multiple PQC signature algorithms will allow the following:
+
+* opt-in migration of wallets and smart contracts to safer cryptographic protocol
+* pay-as-you-go for longer signatures required for post-quantum resistance
+* opt-in resistance to _cryptocalypse_ - surprise breakage of a single major encryption algorithm -
+  by using multiple signature algorithms for redundancy.
+* compliance with US government regulations past July 2023, when all agencies
+  are expected to plan post-quantum migration.
+
+## Specification
+
+[`DSIGN`]() class instance should be provided for the post-quantum signature algorithms.
+
+### Plutus API
+
+Two new builtin functions would be provided:
+
+* A verification function for both pre-quantum and post-quantum resistant
+  signature schemes:
+     - CRYSTALS(DILITHIUM) signatures;
+     - FALCON signatures;
+     - SPHINCS signatures; and
+     - SEQUOA
+* Extensible interface for signature verification functions.
+* A long-term signature verification API useful for smart contract scripts,
+  so that smart contracts do not need to change when new cryptographic protocol
+  is added.
+* Extensible interface for hash verification functions,
+  so that smart contracts do not need to change when hashing algorithm
+  is upgraded.
+
+These would be based on NIST reference implementations,
+as maintained by [`PQClean`](https://github.com/PQClean/PQClean) library.
+This is a reference implementation of the three kinds of signature schemes in C. This 
+implementation would be called from Haskell using direct bindings to C. These
+bindings would be defined in `cardano-base`, using its existing `DSIGN`
+interface, with new builtins in Plutus Core on the basis of the `DSIGN`
+interface for both schemes.
+
+The builtins would be costed as follows: each signature verification has
+linear cost proportional to the number of message bits, and square with respect to key length.
+_Constants will be proportional to the ratio of time as benchmarked against
+Ed25519._
```

Comment:
I think you are suggesting a very complex cost model, namely something that branches on the first argument and then is costed differently (at least wrt constants) depending on that. We could probably add such a thing, but it seems complicated.